### PR TITLE
Parser improvments

### PR DESCRIPTION
**Pull request by brodin**
Added some more values to the parser

### CODE REVIEW

File: parser/src/main/java/spock/traits/PersonTrait.java [hunk 1]
```diff
@@ -12,41 +13,53 @@
...
-
-    Map<String, String> persondata = Splitter.onPattern("\n\\| ")
-        .withKeyValueSeparator('=')
-        .split(textdata);
-
-    for (final Map.Entry<String, String> personEntry : persondata.entrySet()) {
-      output.put(personEntry.getKey(), personEntry.getValue());
+    // Json pointers (https://tools.ietf.org/html/rfc6901)
+    private static final JsonPointer PAGES_POINTER = JsonPointer.compile("/query/pages");
+    private static final JsonPointer REVISION_POINTER = JsonPointer.compile("/revisions/0/*");
```
Comment (luuse) [inline parser/src/main/java/spock/traits/PersonTrait.java]: Remove these and use the extracted ones in the `Utils` class?

Comment (brodin) [inline parser/src/main/java/spock/traits/PersonTrait.java]: Done.

File: parser/src/main/java/spock/traits/Utils.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    private static final JsonPointer REVISION_POINTER = JsonPointer.compile("/revisions/0/*");
+
+    public static String getTextAreaBetween(JsonNode node, Pattern start, Pattern stop) {
+        final String text = getTextInNode(node);
+        final String textTemp = start.split(text)[1];
+
+        return stop.split(textTemp)[0];
+    }
+
+    public static String getTextInNode(JsonNode node) {
```
Comment (luuse) [inline parser/src/main/java/spock/traits/Utils.java]: How about renaming this to `getArticleContents(JsonNode)`? Jackson uses terminology similar to the current method name so it's not apparent what the method does without analyzing the JSON pointers above and the JSON structure that is returned by Wikipedia.

Comment (brodin) [inline parser/src/main/java/spock/traits/Utils.java]: Feel free to change the names! :)

File: parser/src/main/java/spock/traits/CategoryTrait.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    boolean applyThis(JsonNode article, ObjectNode output) {
+        final String articleText = Utils.getTextInNode(article);
+
+        List<String> list = new ArrayList<String>();
+        Matcher matcher = CATEGORY_PATTERN.matcher(articleText);
+        while (matcher.find()) {
+            list.add(matcher.group(1));
+        }
+
+        ObjectMapper mapper = new ObjectMapper();
```
Comment (luuse) [inline parser/src/main/java/spock/traits/CategoryTrait.java]: The `ObjectMapper` object is fairly expensive to create how about adding a third argument to the `Trait#apply` method and friends that accepts the `ObjectMapper` we create in `Fetcher#main` but give to `Parser` which in turn calls `#apply`? This saves us from a few lines in each trait too...

File: parser/src/main/java/spock/traits/InfoBoxTrait.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /**
+     * |image         = President Barack Obama.jpg
+     *
+     * @param key
+     * @return
+     */
+    private String getTerm(String text, String key) {
+        Pattern pattern = Pattern.compile("\\|\\s?(" + key + ")\\s*=\\s(.+?)\\n");
```
Comment (luuse) [inline parser/src/main/java/spock/traits/InfoBoxTrait.java]: Precompiling patterns once can be huge sometimes. Sometimes not so well... might not make a difference in this case.

File: parser/src/test/java/spock/traits/InfoBoxTraitTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
+package spock.traits;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import junit.framework.TestCase;
+
+public class InfoBoxTraitTest extends TestCase {
```
Comment (luuse) [inline parser/src/test/java/spock/traits/InfoBoxTraitTest.java]: Awesome (y)

File: parser/src/main/java/spock/traits/InfoBoxTrait.java (hunk 4 above)
Comment (brodin) [inline parser/src/main/java/spock/traits/InfoBoxTrait.java]: That feels like premature optimization right? When we are getting near the end we could maybe run a profiler instead?

File: parser/src/main/java/spock/traits/CategoryTrait.java (hunk 3 above)
Comment (brodin) [inline parser/src/main/java/spock/traits/CategoryTrait.java]: This makes sense - I also think that we should send the text from the fetcher rather then the entire node, we are only interested in the article as a string anyway.

Comment (luuse) [inline parser/src/main/java/spock/traits/CategoryTrait.java]: The rest of the JSON structure can contain some properties that are easier to parse than the wikitext if you add the right options to the URL we fetch. The categories can be exposed for [example](http://en.wikipedia.org/w/api.php?action=query&format=json&maxlag=5&prop=revisions|categories&titles=Zlatan%20Ibrahimovi%C4%87&rvprop=content).

**Comment by brodin** [comment]
@luuse if you have the time I would love it if you made some of the changes you talked about :)

File: parser/src/main/java/spock/traits/Utils.java (hunk 2 above)
Comment (brodin) [inline parser/src/main/java/spock/traits/Utils.java]: Done.

File: parser/src/main/java/spock/traits/CategoryTrait.java (hunk 3 above)
Comment (brodin) [inline parser/src/main/java/spock/traits/CategoryTrait.java]: Sure! :)

**Comment by brodin** [comment]
I found a problem with the `getTerm` regexp, it will match the following wrong

```
| birth_place     = [[Allentown, Pennsylvania]], [[U.S.]]
| death_date      =
| death_place     =
| abc     = 123

```

When doing `getTerm(text, "death_date")` it will match all the way down to death_place, I am no pro at regexp, can you look at it @luuse ?

**Comment by luuse** [comment]
This PR is getting rather sprawly, how about just merging this and create issues for the remaining things? Hard to do anything without risking conflicts with this atm.

**Comment by brodin** [comment]
Fair enough, go ahead.

2015-05-20 9:26 GMT+02:00 Andreas Falk <notifications@github.com>:

**Pull request merged**